File: src/net/dinkla/raytracer/math/Point2D.scala
package net.dinkla.raytracer.math

/**
 * Created by Dinkla on 10.07.2015.
 */

abstract class Point2D[@specialized(Float, Double) F](x: F, y: F)(implicit numType: Fractional[F]) extends Element2D[F](x, y)  {

  def unary_-(): Point2D[F]           // unary minus

  def +(v: Vector2D[F]): Point2D[F]        = plus(v)     // Point plus Vector gives a Point
  def +(p: Point2D[F]): Vector2D[F]        = plus(p)     // Point plus Point gives a Vector

  def plus(v: Vector2D[F]): Point2D[F]     // Point plus Vector gives a Point
  def plus(p: Point2D[F]): Vector2D[F]     // Point plus Point gives a Vector

  def -(v: Vector2D[F]): Point2D[F]        = plus(-v)    // Point minus Vector gives a Point
  def -(p: Point2D[F]): Vector2D[F]        = plus(-p)    // Point minus Point gives a Vector

  def minus(v: Vector2D[F]): Point2D[F]    = plus(-v)    // Point minus Vector gives a Point
  def minus(p: Point2D[F]): Vector2D[F]    = plus(-p)    // Point minus Point gives a Vector

  def toVector: Vector2D[F]

}

object Point2D {

  def Point2D(x: Double, y: Double) = new Point2DD(x, y)
  def Point2D(x: Float, y: Float) = new Point2DF(x, y)

}

File: src/commonMain/kotlin/net/dinkla/raytracer/math/Point2D.kt
package net.dinkla.raytracer.math

import kotlin.math.sqrt

data class Point2D(val x: Double, val y: Double) {

    operator fun unaryMinus() = Vector2D(-x, -y)

    operator fun plus(v: Vector2D) = Point2D(x + v.x, y + v.y)

    operator fun minus(v: Vector2D) = Point2D(x - v.x, y - v.y)

    val length: Double
        get() = sqrt(x * x + y * y)

    override fun toString(): String = "($x,$y)"

    companion object {
        val ORIGIN = Point2D(0.0, 0.0)
    }
}
